[TextureMapper] Fix preserve-3D intersection rendering

zhani opened the pull request:
#### d40d59e4e5506747ab095ac2c3b337679e38265e
<pre>
[TextureMapper] Fix preserve-3D intersection rendering
<a href="https://bugs.webkit.org/show_bug.cgi?id=282682">https://bugs.webkit.org/show_bug.cgi?id=282682</a>

Reviewed by Fujii Hironori.

Preserve-3d does not correctly render intersections, and also applying
filters in a 3D scene can break the rendering order. This patch addresses
these issues by implementing the composition and painting logic from the
CSS Transforms Module Level 2.

In a 3D rendering context, rendering and sorting of elements is done as
follows:

1. The element establishing the 3D rendering context, and each other 3D
transformed element participating in the 3D rendering context, is rendered
into its own plane. This plane includes the element’s backgrounds, borders,
other box decorations, content, and descendant elements, excluding any
descendant elements that have their own plane (and their descendants).
This rendering is done according to CSS 2.1,Appendix E, Section E.2 Painting Order.

2. Intersection is performed between this set of planes, according to
Newell’s algorithm, with the planes transformed by the accumulated 3D
transformation matrix. Coplanar 3D transformed elements are rendered in
painting order.

* Source/WebCore/platform/TextureMapper.cmake:
* Source/WebCore/platform/graphics/texmap/FloatPlane3D.cpp: Added.
(WebCore::FloatPlane3D::FloatPlane3D):
* Source/WebCore/platform/graphics/texmap/FloatPlane3D.h: Added.
(WebCore::FloatPlane3D::normal const):
(WebCore::FloatPlane3D::distanceConstant const):
(WebCore::FloatPlane3D::distanceToPoint const):
* Source/WebCore/platform/graphics/texmap/FloatPolygon3D.cpp: Added.
(WebCore::FloatPolygon3D::FloatPolygon3D):
(WebCore::FloatPolygon3D::split const):
* Source/WebCore/platform/graphics/texmap/FloatPolygon3D.h: Added.
(WebCore::FloatPolygon3D::vertexAt const):
(WebCore::FloatPolygon3D::numberOfVertices const):
(WebCore::FloatPolygon3D::normal const):
* Source/WebCore/platform/graphics/texmap/TextureMapper.cpp:
(WebCore::TextureMapper::beginClip):
* Source/WebCore/platform/graphics/texmap/TextureMapper.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperLayer::computeTransformsRecursive):
(WebCore::TextureMapperLayer::paintSelf):
(WebCore::TextureMapperLayer::paintWith3DRenderingContext):
(WebCore::TextureMapperLayer::collect3DSceneLayers):
(WebCore::TextureMapperLayer::effectiveLayerRect const):
(WebCore::TextureMapperLayer::paintPreserves3DHolePunch): Deleted.
(WebCore::TextureMapperLayer::sortByZOrder): Deleted.
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
(WebCore::TextureMapperLayer::toSurfaceTransform const):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp: Added.
(WebCore::TextureMapperLayer3DRenderingContext::paint):
(WebCore::TextureMapperLayer3DRenderingContext::buildTree):
(WebCore::TextureMapperLayer3DRenderingContext::traverseTreeAndPaint):
(WebCore::TextureMapperLayer3DRenderingContext::classifyPolygon):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.h: Added.
* Source/WebCore/platform/graphics/transforms/TransformationMatrix.cpp:
(WebCore::TransformationMatrix::transpose const):
* Source/WebCore/platform/graphics/transforms/TransformationMatrix.h:

Canonical link: <a href="https://commits.webkit.org/286605@main">https://commits.webkit.org/286605@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6fd2120e5b2cd16166a771365861101e940eed78

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76338 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55370 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29240 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80949 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27610 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78454 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64512 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3664 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59921 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17982 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79405 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49758 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65558 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40250 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47156 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23045 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25932 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68283 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23379 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82305 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3710 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2439 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68087 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3864 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65530 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67395 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11354 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9456 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11834 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3658 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6465 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3681 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7110 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/5439 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fujii commented on Source/WebCore/platform/graphics/FloatPlane3D.h:
```diff
@@ -0,0 +1,55 @@
+
+private:
+    FloatPoint3D m_normal;
+    float m_d;
```
`m_d` seems to violate the WebKit coding style guildeline.
https://webkit.org/code-style-guidelines/#names-full-words

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/FloatPolygon3D.h:
```diff
@@ -0,0 +1,56 @@
+#include "FloatPlane3D.h"
+#include "FloatPoint3D.h"
+#include "FloatRect.h"
+#include "transforms/TransformationMatrix.h"
```
#include "TransformationMatrix.h"

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/FloatPolygon3D.cpp:
```diff
@@ -0,0 +1,114 @@
+            positiveVertices.append(vi);
+
+            if (dj < -epsilon) {
+
```
I think WebKit doesn't put a blank line here.

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/FloatPolygon3D.cpp:
```diff
@@ -0,0 +1,114 @@
+    FloatPolygon3D negativePolygon;
+    negativePolygon.m_vertices = WTFMove(negativeVertices);
+
+    return std::make_pair(negativePolygon, positivePolygon);
```
return { negativePolygon, positivePolygon};

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapper.cpp:
```diff
@@ -1320,6 +1320,68 @@ void TextureMapper::beginClip(const TransformationMatrix& modelViewMatrix, const
+
+    unsigned numberOfVertices = polygon.numberOfVertices();
+    Vector<GLfloat> polygonVertices;
+    polygonVertices.reserveCapacity(numberOfVertices*2);
```
Insert spaces. `numberOfVertices * 2`.
https://webkit.org/code-style-guidelines/#spacing-binary-ternary-op

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -0,0 +1,144 @@
+
+    // if polygon is facing away from camera then swap nodes to reverse
+    // the traversal order
+    if (plane.normal().z() < 0.f)
```
`0`
https://webkit.org/code-style-guidelines/#float-suffixes

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -1047,64 +1001,38 @@ void TextureMapperLayer::paintFlattened(TextureMapperPaintOptions& options)
-            for (int y = rect.y(); y < rect.maxY(); y += maxTextureSize.height()) {
-                IntRect tileRect(IntPoint(x, y), maxTextureSize);
-                tileRect.intersect(rect);
-                auto surface = options.textureMapper.acquireTextureFromPool(tileRect.size(), { BitmapTexture::Flags::SupportsAlpha, BitmapTexture::Flags::DepthBuffer });
```
We can remove `Flags::DepthBuffer`.

zhani replied:
Intermediate surface is removed completely

fujii replied:
I mean we can remove `Flags::DepthBuffer` flag.

zhani replied:
I'm not sure if I understand what you mean. Patch removes these lines. Do you mean that these lines should not be removed and only Flags::DepthBuffer removed?

fujii replied:
We no longer use Flags::DepthBuffer flag after this change. Let's remove Flags::DepthBuffer flag too. It's OK to split a patch if you want.

zhani replied:
Remove Flags::DepthBuffer from were? This patch has already removed Flags::DepthBuffer usage from TextureMapperLayer.

Do you mean that we should remove DepthBuffer enumeration value from BitmapTexture completely?

zhani replied:
oh, now I understand. I'm very surprised that this was the only usage for that flag in whole codebase

zhani replied:
Maybe it's better to do follow-up patch to remove it so that in case it's needed again later on by some other code then it's easy to revert, what do you think?

fujii replied:
Sounds good.

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.h:
```diff
@@ -0,0 +1,97 @@
+                }
+            }
+
+            return FloatPolygon(WTFMove(vertices), WindRule::EvenOdd);
```
I think WebKit writes this as `return { WTFMove(vertices), WindRule::EvenOdd) };`.

fujii replied:
`WindRule::EvenOdd` is not used and confusing.
I think you don't need to use `FloatPolygon` here.
I think `Vector<FloatPoint>` is enough. `using ClipPolygon = Vector<FloatPoint>;` .

zhani replied:
I decided to use FloatPolygon because it was handy already existing class and I think it clearly shows that we go to 2D coordintes. WindRule is not used but it's mandatory parameter in constructor. Actually it should be WindRule::NonZero as purpose of this polygon is clipArea. I changed it to NonZero. I don't have strong opinion here whether to use Vector<FloatPoint> or FloatPolygon

zhani replied:
Actually after second though maybe it's better to go with `Vector<FloatPoint>`. `FloatPolygon` constructor implementation does unnecessary things which we don't need.

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapper.cpp:
```diff
@@ -1320,6 +1320,68 @@ void TextureMapper::beginClip(const TransformationMatrix& modelViewMatrix, const
     clipStack().applyIfNeeded();
 }
 
+void TextureMapper::beginClip(const TransformationMatrix& modelViewMatrix, const FloatPolygon& polygon)
```
TextureMapper is doing antialiasing for some layers with `TextureMapperShaderProgram::Antialiasing`.
Does it work well with this clipping?

fujii replied:
It extends edges 1px.
https://github.com/WebKit/WebKit/blob/5da68fa7157872ef250bf4f159e14f4a2e3d58f1/Source/WebCore/platform/graphics/texmap/TextureMapperShaderProgram.cpp#L83

zhani replied:
I think it works. I need to check more closely the stencil usage. Most of this beginClip code is copy pasted from existing beginClip for FloatRect function

zhani replied:
@fujii you are right, this clipping approach currently removes antialiasing, so it requires adjustment. To achieve a pixel-perfect stencil clip for the antialiased parts, we should render the stencil clip using the same shader as layer rendering. However, that shader is incompatible with how polygon clips must be rendered. Additionally, the intersection edge should remain un-antialiased, so we need selective antialiasing along the edge. I'm evaluating a few approaches to refactor the antialiasing to address these requirements.

fujii replied:
I don't mind disabling antialiasing in 3d rendering contexts if it's difficult.

zhani replied:
Currently, the antialiasing logic is designed specifically for rectangles, expecting that they fall within bounds 0..1 and are mapped to the correct size through the transformation matrix. Although the rectangle is divided into triangles, the expectation remains that these triangles stay within the rectangle shape. I’ve also observed cases where antialiasing interpolates inward, causing the edge to fade in an unintended direction.

Adapting the current logic to handle arbitrary polygons, which is needed for accurate clipping, seems quite challenging (if not impossible) as it stands. I’m implementing a new antialiasing shader that will work with arbitrary polygons but it will take an additional week or two to have it ready for review.

It might make sense to land this patch first and handle antialiasing in a separate patch, what do you think? The shader update could introduce layout test issues, so separating them probably makes it easier to review and maintain focus and clarity on correct issue.

zhani replied:
With this patch antialiasing doesn't completely fail in 3D. Non intersecting layers will have full antialiasing and also intersecting layers have some antialiasing but depending on transformation and perspective some of the antialiasing might be clipped off because stencil clip cannot use antialiasing shader logic. But in most of the cases it's hard to see the issue without going into pixel level detail because there's still some antialiasing effect present even if clipped

fujii replied:
Sounds nice.

fujii commented on Source/WebCore/Sources.txt:
```diff
@@ -2403,9 +2403,11 @@ platform/graphics/DisplayRefreshMonitorClient.cpp
 platform/graphics/FloatPoint.cpp
 platform/graphics/FloatPoint3D.cpp
 platform/graphics/FloatPolygon.cpp
+platform/graphics/FloatPolygon3D.cpp
```
FloatPlane3D and FloatPolygon3D are used only by TextureMapper. Cocoa port won't use it because it uses CoreAnimation. I think these files should be moved to platform/graphics/texmap.
Or, enclosed .cpp and .h by `#if USE(TEXTURE_MAPPER)` to express that this is only for TextureMapper.

zhani replied:
Moved to platform/graphics/texmap

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/FloatPlane3D.cpp:
```diff
@@ -0,0 +1,43 @@
+FloatPlane3D::FloatPlane3D(const FloatPoint3D& normal, const FloatPoint3D& point)
+    : m_normal(normal)
+{
+    m_distanceConstant = m_normal.dot(point);
```
```suggestion
    : m_normal(normal)
    , m_distanceConstant(m_normal.dot(point))
{
```

zhani replied:
Fixed

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/FloatPolygon3D.cpp:
```diff
@@ -0,0 +1,124 @@
+    m_vertices.append(transform.mapPoint(FloatPoint3D(rect.minXMaxYCorner())));
+
+    auto inverse = transform.inverse();
+    if (inverse) {
```
```suggestion
    if (auto inverse = transform.inverse()) {
```

zhani replied:
Fixed

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/FloatPolygon3D.cpp:
```diff
@@ -0,0 +1,124 @@
+
+    auto inverse = transform.inverse();
+    if (inverse) {
+        m_normal = (*inverse).transpose().mapPoint(m_normal);
```
```suggestion
        m_normal = inverse->transpose().mapPoint(m_normal);
```

zhani replied:
Fixed

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/FloatPolygon3D.cpp:
```diff
@@ -0,0 +1,124 @@
+    auto inverse = transform.inverse();
+    if (inverse) {
+        m_normal = (*inverse).transpose().mapPoint(m_normal);
+        m_normal.normalize();
```
This is done in both cases, it could be moved after the if and removed from both branches

zhani replied:
Fixed

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/FloatPolygon3D.cpp:
```diff
@@ -0,0 +1,124 @@
+    Vector<FloatPoint3D> positiveVertices;
+    Vector<FloatPoint3D> negativeVertices;
+
+    const float epsilon = 1e-6f; // Tolerance for floating point comparisons
```
`std::numeric_limits<float>::epsilon()` ?

zhani replied:
At some point I had some issues std::numeric_limits<float>::epsilon() being too small and causing artifacts but it was during the initial development of the patch. I tested again with std::numeric_limits<float>::epsilon() and cannot see issues anymore. So I applied this now

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/FloatPolygon3D.cpp:
```diff
@@ -0,0 +1,124 @@
+
+    FloatPolygon3D negativePolygon;
+    negativePolygon.m_vertices = WTFMove(negativeVertices);
+    negativePolygon.m_normal = m_normal;
```
I guess you could add a private FloatPolygon3D constructor that receives the vertices and point

zhani replied:
Fixed

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
```diff
@@ -82,6 +82,7 @@ class TextureMapperLayer : public CanMakeWeakPtr<TextureMapperLayer> {
     FloatSize size() const { return m_state.size; }
     float opacity() const { return m_state.opacity; }
     TransformationMatrix transform() const { return m_state.transform; }
+    TransformationMatrix toSurfaceTransform() const { return m_layerTransforms.combined; }
```
Do we need to copy? or can we return a `const TransformationMatrix&` ?

zhani replied:
Fixed

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -0,0 +1,145 @@
+// Build BSP tree for rendering polygons with painter's algorithm.
+// For more information:
+// https://en.wikipedia.org/wiki/Binary_space_partitioning
+void TextureMapperLayer3DRenderingContext::buildTree(TextureMapperLayerNode* root, Deque<TextureMapperLayerPolygon>& polygons)
```
```suggestion
void TextureMapperLayer3DRenderingContext::buildTree(TextureMapperLayerNode& root, Deque<TextureMapperLayerPolygon>& polygons)
```

zhani replied:
Fixed

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -0,0 +1,145 @@
+        layerList.append({ { layer->effectiveLayerRect(), layer->toSurfaceTransform() }, layer, false });
+
+    auto root = makeUnique<TextureMapperLayerNode>(layerList.takeFirst());
+    buildTree(root.get(), layerList);
```
```suggestion
    buildTree(*root, layerList);
```

zhani replied:
Fixed

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -0,0 +1,145 @@
+    if (polygons.isEmpty())
+        return;
+
+    auto rootGeometry = root->firstPolygon().geometry;
```
```suggestion
    auto& rootGeometry = root.firstPolygon().geometry;
```

zhani replied:
Fixed

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -0,0 +1,145 @@
+    Deque<TextureMapperLayerPolygon> backList, frontList;
+    for (auto& polygon : polygons) {
+        auto polygonPosition = classifyPolygon(polygon, rootPlane);
+        switch (polygonPosition) {
```
```suggestion
        switch (classifyPolygon(polygon, rootPlane)) {
```

zhani replied:
Fixed

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -0,0 +1,145 @@
+
+    if (!frontList.isEmpty()) {
+        root->frontNode = makeUnique<TextureMapperLayerNode>(frontList.takeFirst());
+        buildTree(root->frontNode.get(), frontList);
```
```suggestion
        buildTree(*root.frontNode, frontList);
```

zhani replied:
Fixed

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -0,0 +1,145 @@
+
+    if (!backList.isEmpty()) {
+        root->backNode = makeUnique<TextureMapperLayerNode>(backList.takeFirst());
+        buildTree(root->backNode.get(), backList);
```
```suggestion
        buildTree(*root.backNode, backList);
```

zhani replied:
Fixed

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -0,0 +1,145 @@
+
+void TextureMapperLayer3DRenderingContext::traverseTreeAndPaint(TextureMapperLayerNode* node, const std::function<void(TextureMapperLayer*, const FloatPolygon&)>& paintLayerFunction)
+{
+    auto geometry = node->firstPolygon().geometry;
```
```suggestion
    auto& geometry = node.firstPolygon().geometry;
```

zhani replied:
Fixed

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -0,0 +1,145 @@
+    }
+}
+
+void TextureMapperLayer3DRenderingContext::traverseTreeAndPaint(TextureMapperLayerNode* node, const std::function<void(TextureMapperLayer*, const FloatPolygon&)>& paintLayerFunction)
```
```suggestion
void TextureMapperLayer3DRenderingContext::traverseTreeAndPaint(TextureMapperLayerNode& node, const std::function<void(TextureMapperLayer*, const FloatPolygon&)>& paintLayerFunction)
```

zhani replied:
Fixed

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -0,0 +1,145 @@
+        std::swap(frontNode, backNode);
+
+    if (backNode)
+        traverseTreeAndPaint(backNode, paintLayerFunction);
```
```suggestion
        traverseTreeAndPaint(*backNode, paintLayerFunction);
```

zhani replied:
Fixed

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -0,0 +1,145 @@
+    FloatPlane3D plane(geometry.normal(), geometry.vertexAt(0));
+
+    auto* frontNode = node->frontNode.get();
+    auto* backNode = node->backNode.get();
```
```suggestion
    auto* frontNode = node.frontNode.get();
    auto* backNode = node.backNode.get();
```

zhani replied:
Fixed

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -0,0 +1,145 @@
+    if (backNode)
+        traverseTreeAndPaint(backNode, paintLayerFunction);
+
+    for (auto& polygon : node->polygons)
```
```suggestion
    for (auto& polygon : node.polygons)
```

zhani replied:
Fixed

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -0,0 +1,145 @@
+        paintLayerFunction(polygon.layer, polygon.layerClipArea());
+
+    if (frontNode)
+        traverseTreeAndPaint(frontNode, paintLayerFunction);
```
```suggestion
        traverseTreeAndPaint(*frontNode, paintLayerFunction);
```

zhani replied:
Fixed

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    if (!frontList.isEmpty()) {
+        root->frontNode = makeUnique<TextureMapperLayerNode>(frontList.takeFirst());
```
```suggestion
        root.frontNode = makeUnique<TextureMapperLayerNode>(frontList.takeFirst());
```

zhani replied:
Fixed

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    if (!backList.isEmpty()) {
+        root->backNode = makeUnique<TextureMapperLayerNode>(backList.takeFirst());
```
```suggestion
        root.backNode = makeUnique<TextureMapperLayerNode>(backList.takeFirst());
```

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.h:
```diff
@@ -0,0 +1,98 @@
+
+    struct TextureMapperLayerPolygon final {
+        FloatPolygon3D geometry;
+        TextureMapperLayer* layer;
```
```suggestion
        TextureMapperLayer* layer { nullptr };
```

zhani replied:
Fixed

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.h:
```diff
@@ -0,0 +1,98 @@
+    struct TextureMapperLayerPolygon final {
+        FloatPolygon3D geometry;
+        TextureMapperLayer* layer;
+        bool isSplitted;
```
```suggestion
        bool isSplitted { false };
```

zhani replied:
Fixed

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.h:
```diff
@@ -0,0 +1,98 @@
+            vertices.reserveCapacity(numVertices);
+            auto inverse = layer->toSurfaceTransform().inverse();
+            if (isSplitted && inverse) {
+                auto transformToLayer = *inverse;
```
Do we need to copy this? You can use inverse->mapPoint below, no?

zhani replied:
Fixed

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.h:
```diff
@@ -0,0 +1,98 @@
+        TextureMapperLayer* layer;
+        bool isSplitted;
+
+        FloatPolygon layerClipArea() const
```
Move the function before the variables block.

zhani replied:
Fixed

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.h:
```diff
@@ -0,0 +1,98 @@
+    struct TextureMapperLayerNode final {
+        WTF_MAKE_FAST_ALLOCATED;
+
+    public:
```
I don't this is needed for a struct.

zhani replied:
Fixed

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.h:
```diff
@@ -0,0 +1,98 @@
+    };
+
+    struct TextureMapperLayerNode final {
+        WTF_MAKE_FAST_ALLOCATED;
```
```suggestion
        WTF_MAKE_STRUCT_FAST_ALLOCATED;
```

zhani replied:
Fixedr

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.h:
```diff
@@ -0,0 +1,98 @@
+        WTF_MAKE_FAST_ALLOCATED;
+
+    public:
+        TextureMapperLayerNode(TextureMapperLayerPolygon&& polygon)
```
```suggestion
        explicit TextureMapperLayerNode(TextureMapperLayerPolygon&& polygon)
```

zhani replied:
Fixed

carlosgcampos reviewed (commented):
Thanks!

fujii commented on Source/WebCore/platform/graphics/texmap/FloatPlane3D.h:
```diff
@@ -0,0 +1,58 @@
+    // Getter for the distance from the origin (plane constant d)
+    float distanceConstant() const { return m_distanceConstant; }
+
+    // Signed distance. The is of the return value is positive
```
Do you need "The is of" here?

zhani replied:
Typo. It should be "The sign of". Fixed

fujii reviewed (approved):
Great improvement! Looks good to me. Thank you very much.

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/FloatPlane3D.h:
```diff
@@ -0,0 +1,58 @@
+
+private:
+    FloatPoint3D m_normal;
+    float m_distanceConstant;
```
I'd add { 0.0f } to be safe, if for some reasons different constructors are added at some point, that skip setting m_distanceConstant.

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/FloatPolygon3D.cpp:
```diff
@@ -0,0 +1,119 @@
+
+FloatPolygon3D::FloatPolygon3D(const FloatRect& rect, const TransformationMatrix& transform)
+{
+    m_vertices.append(transform.mapPoint(FloatPoint3D(rect.minXMinYCorner())));
```
Either reserve capacity to `4` or add an inline capacity to the m_vertices vector, Vector<FloatPoint3D, 4>.

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/FloatPolygon3D.cpp:
```diff
@@ -0,0 +1,119 @@
+// - Construct two new polygons from the collected vertices.
+std::pair<FloatPolygon3D, FloatPolygon3D> FloatPolygon3D::split(const FloatPlane3D& plane) const
+{
+    Vector<FloatPoint3D> positiveVertices;
```
You could reverseCapacity to m_vertices.size() here, and shrinkToFit() in the end.

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/FloatPolygon3D.cpp:
```diff
@@ -0,0 +1,119 @@
+        float di = plane.distanceToPoint(vi);
+        float dj = plane.distanceToPoint(vj);
+
+        bool viPos = di > epsilon;
```
I'd avoid those two local variables, they are only used in the following if conditions. Makes it a bit easier to read IMHO.

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/FloatPolygon3D.h:
```diff
@@ -0,0 +1,61 @@
+    FloatPolygon3D() = default;
+    FloatPolygon3D(const FloatRect&, const TransformationMatrix&);
+
+    const FloatPoint3D& vertexAt(unsigned index) const { return m_vertices[index]; }
```
Better use .at(index) here, to enable overflow checking.

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/FloatPolygon3D.h:
```diff
@@ -0,0 +1,61 @@
+    FloatPolygon3D(Vector<FloatPoint3D>&&, const FloatPoint3D&);
+
+    Vector<FloatPoint3D> m_vertices;
+    FloatPoint3D m_normal = { 0, 0, 1 };
```
Omit the '='.

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapper.cpp:
```diff
@@ -1320,6 +1321,68 @@ void TextureMapper::beginClip(const TransformationMatrix& modelViewMatrix, const
+    clipStack().push();
+    data().initializeStencil();
+
+    Ref<TextureMapperShaderProgram> program = data().getShaderProgram(TextureMapperShaderProgram::SolidColor);
```
auto

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -0,0 +1,144 @@
+    if (layers.isEmpty())
+        return;
+
+    Deque<TextureMapperLayerPolygon> layerList;
```
Maybe provide a reasonable inline capacity here, to avoid heap allocations?

zhani replied:
It's impossible to tell good number. It can be anything from 1 to ...

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -0,0 +1,144 @@
+    auto& rootGeometry = root.firstPolygon().geometry;
+    FloatPlane3D rootPlane(rootGeometry.normal(), rootGeometry.vertexAt(0));
+
+    Deque<TextureMapperLayerPolygon> backList, frontList;
```
Two lines per WebKit style.

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -0,0 +1,144 @@
+        float distance = plane.distanceToPoint(vertex);
+
+        if (distance > epsilon)
+            inFrontCount++;
```
s/foo++/++foo/

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -0,0 +1,144 @@
+
+    int inFrontCount = 0;
+    int behindCount = 0;
+    for (unsigned i = 0; i < polygon.geometry.numberOfVertices(); i++) {
```
s/i++/++i/

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.h:
```diff
@@ -0,0 +1,96 @@
+            vertices.reserveCapacity(numVertices);
+            auto toLayerTransform = layer->toSurfaceTransform().inverse();
+            if (isSplitted && toLayerTransform) {
+                for (unsigned i = 0; i < numVertices; i++) {
```
s/i++/++i/

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.h:
```diff
@@ -0,0 +1,96 @@
+            if (isSplitted && toLayerTransform) {
+                for (unsigned i = 0; i < numVertices; i++) {
+                    auto v = toLayerTransform->mapPoint(geometry.vertexAt(i));
+                    vertices.append(FloatPoint(v.x(), v.y()));
```
append({ v.x(), v.y() }) ?

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.h:
```diff
@@ -0,0 +1,96 @@
+        }
+
+        FloatPolygon3D geometry;
+        TextureMapperLayer* layer = { nullptr };
```
Omit the '='.

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.h:
```diff
@@ -0,0 +1,96 @@
+
+        FloatPolygon3D geometry;
+        TextureMapperLayer* layer = { nullptr };
+        bool isSplitted = { false };
```
Ditto.

zhani replied:
Fixed

nikolaszimmermann reviewed (commented):
Looks really nice! Found a few small issues - would be nice if you could still resolve them.

zhani committed:
[TextureMapper] Fix preserve-3D intersection rendering
https://bugs.webkit.org/show_bug.cgi?id=282682

Reviewed by Fujii Hironori.

Preserve-3d does not correctly render intersections, and also applying
filters in a 3D scene can break the rendering order. This patch addresses
these issues by implementing the composition and painting logic from the
CSS Transforms Module Level 2.

In a 3D rendering context, rendering and sorting of elements is done as
follows:

1. The element establishing the 3D rendering context, and each other 3D
transformed element participating in the 3D rendering context, is rendered
into its own plane. This plane includes the element’s backgrounds, borders,
other box decorations, content, and descendant elements, excluding any
descendant elements that have their own plane (and their descendants).
This rendering is done according to CSS 2.1,Appendix E, Section E.2 Painting Order.

2. Intersection is performed between this set of planes, according to
Newell’s algorithm, with the planes transformed by the accumulated 3D
transformation matrix. Coplanar 3D transformed elements are rendered in
painting order.

* Source/WebCore/platform/TextureMapper.cmake:
* Source/WebCore/platform/graphics/texmap/FloatPlane3D.cpp: Added.
(WebCore::FloatPlane3D::FloatPlane3D):
* Source/WebCore/platform/graphics/texmap/FloatPlane3D.h: Added.
(WebCore::FloatPlane3D::normal const):
(WebCore::FloatPlane3D::distanceConstant const):
(WebCore::FloatPlane3D::distanceToPoint const):
* Source/WebCore/platform/graphics/texmap/FloatPolygon3D.cpp: Added.
(WebCore::FloatPolygon3D::FloatPolygon3D):
(WebCore::FloatPolygon3D::split const):
* Source/WebCore/platform/graphics/texmap/FloatPolygon3D.h: Added.
(WebCore::FloatPolygon3D::vertexAt const):
(WebCore::FloatPolygon3D::numberOfVertices const):
(WebCore::FloatPolygon3D::normal const):
* Source/WebCore/platform/graphics/texmap/TextureMapper.cpp:
(WebCore::TextureMapper::beginClip):
* Source/WebCore/platform/graphics/texmap/TextureMapper.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperLayer::computeTransformsRecursive):
(WebCore::TextureMapperLayer::paintSelf):
(WebCore::TextureMapperLayer::paintWith3DRenderingContext):
(WebCore::TextureMapperLayer::collect3DSceneLayers):
(WebCore::TextureMapperLayer::effectiveLayerRect const):
(WebCore::TextureMapperLayer::paintPreserves3DHolePunch): Deleted.
(WebCore::TextureMapperLayer::sortByZOrder): Deleted.
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
(WebCore::TextureMapperLayer::toSurfaceTransform const):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp: Added.
(WebCore::TextureMapperLayer3DRenderingContext::paint):
(WebCore::TextureMapperLayer3DRenderingContext::buildTree):
(WebCore::TextureMapperLayer3DRenderingContext::traverseTreeAndPaint):
(WebCore::TextureMapperLayer3DRenderingContext::classifyPolygon):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.h: Added.
* Source/WebCore/platform/graphics/transforms/TransformationMatrix.cpp:
(WebCore::TransformationMatrix::transpose const):
* Source/WebCore/platform/graphics/transforms/TransformationMatrix.h:

Canonical link: https://commits.webkit.org/286605@main
